include/boost/decimal/detail/cmath/log10.hpp

```diff
@@ -26,21 +27,102 @@ template <typename T>
 constexpr auto log10_impl(T x) noexcept
     BOOST_DECIMAL_REQUIRES(detail::is_decimal_floating_point_v, T)
 {
-    // TODO(ckormanyos) Actually this is eintirely preliminary. The implementation
-    // of log10 will/should-be entirely re-worked with specific base-10-relevant details.
-
-    // TODO(ckormanyos) Handle non-normal varguments.
-
-    // TODO(ckormanyos) Put in a basic check for pure powers of 10, resulting
-    // in an exact result.
-
-    // Starting point: See implementation of log() and adapt to the following series.
-    // Series[Log[10] Log[10, (1 + (z/2))/(1 - (z/2))], {z, 0, 17}]
-
-    return log(x) / numbers::ln10_v<T>;
+    constexpr T zero { 0, 0 };
+    constexpr T one  { 1, 0 };
+
+    T result { };
+
+    if (isnan(x))
+    {
+        result = x;
+    }
+    else if (isinf(x))
+    {
+        result = (!signbit(x)) ? x: std::numeric_limits<T>::quiet_NaN();
+    }
+    else if (x < one)
+    {
+        // Handle reflection, the [+/-] zero-pole, and non-pole, negative x.
+        if (x > zero)
+        {
+            result = -log10(one / x);
+        }
+        else if ((x == zero) || (-x == zero))
+        {
+            // Actually, this should be equivalent to -HUGE_VAL.
+
+            result = -std::numeric_limits<T>::infinity();
+        }
+        else
+        {
+            result = std::numeric_limits<T>::quiet_NaN();
+        }
+    }
+    else if(x > one)
+    {
+        // The algorithm for base-10 logarithm is based on Chapter 5, pages 35-36
+        // of Cody and Waite, Software Manual for the Elementary Functions,
+        // Prentice Hall, 1980.
+
+        int exp10val { };
+
+        T g { frexp10(x, &exp10val) };
+
+        while(g > one)
+        {
+            // TODO(ckormanyos) Do we really have to multiply individual powers
+            // of 10 here? Or is there a reliable way to simply use pow10()?
+            // For instance count intagral orders of exp10val and divide by pow10().
```

Comment:
The significand is normalized before being returned by frexp10 so you can divide by pow10(digits10 - 1) to avoid the loop

Reply:
>  significand is normalized before being returned by frexp10 so you can divide by pow10(digits10 - 1) to avoid the loop

OK @mborland cool but now I'm stuck on another point. Do you know how to recognize a pure power of $10$ without burning too much CPU cycling?

Reply:
You could use `detail::remove_trailing_zeros` and compare to 1. That is about as efficient as it's going to get I think.